File: 19_数学/因数筛/getFactorsAll.py
from typing import List


def getFactorsAll(max: int) -> List[List[int]]:
    """预处理 1~max 的所有数的因数."""
    res = [[] for _ in range(max + 1)]
    for f in range(1, max + 1):
        for m in range(f, max + 1, f):
            res[m].append(f)
    return res


assert getFactorsAll(10) == [  # noqa
    [],
    [1],
    [1, 2],
    [1, 3],
    [1, 2, 4],
    [1, 5],
    [1, 2, 3, 6],
    [1, 7],
    [1, 2, 4, 8],
    [1, 3, 9],
    [1, 2, 5, 10],
]

File: 19_数学/因数筛/getFactorsAll.go
// 预处理多个数的因数/GetAllFactors

package main

import "fmt"

func main() {
	fmt.Println(GetFactorsAll(10))
}

// 预处理 1~max 的所有数的因数.
func GetFactorsAll(max int32) (res [][]int32) {
	res = make([][]int32, max+1)
	for f := int32(1); f <= max; f++ {
		for m := f; m <= max; m += f {
			res[m] = append(res[m], f)
		}
	}
	return
}
